proposals/025-lambda-1.md

```diff
@@ -0,0 +1,224 @@
+
+## Proposal
+
+The proposal is to add the following syntax for lambda expressions:
+```slang
+(parameter_list) => expression
+```
+
+or
+
+```slang
+(parameter_list) => { statement_list }
+```
+
+Where `parameter_list` is a comma-separated list of parameters, same as those in ordinary functions,
+and `expression` or `statement_list` defines the body of the lambda. For examples, these two lambdas
+achieve similar results:
+
+```slang
+(int x) => return x > 0 ? x : 0
+
+(int x) => {
+    if (x > 0) {
+        return x;
+    } else {
+        return 0;
+    }
+}
+```
+
+A lambda expression will evaluate to an annoymous struct type that implements the `IFunc` interface
+during type checking. The return type of the lambda function is determined by the body expression (in the case
+of the lambda expression contains a simple expression body), or the value of the return statements in the
+case of a statement body. If the lambda function body contains more than one return statements, then the return
+values from all return statements must be exactly the same.
+
+In the future, we will also extend lambda expressions to allow them to conform to other interfaces including
+`IDifferentiableFunc` or `IMutatingFunc`.
+
+Lambda expressions can be used in positions that accepts an `IFunc`:
+
+```
+void apply(IFunc<int, float> x) {...}
+
+void test()
+{
+    apply((float x)=>(int)x+1); // OK, passing lambda to `IFunc<int, float>`.
+}
+```
+
+## Translation
+
+Immutable lambda expressions translates into a struct type implementing the corresponding `IFunc` interface.
+For example, given the following code:
+
+```slang
+void test()
+{
+    int c = 0;
+    let lam = (int x) => x + c;
+    int d = lam(2);
+}
+```
+
+The compiler will translate it into:
+
+```slang
+void test()
+{
+    int c = 0;
+    struct _slang_Lambda_test_0 : IFunc<int, int> {
+        int c;
+        __init(int in_c) {
+            c = in_c;
+        }
+        int operator()(int x) {
+            return x + c;
+        }
+    }
+    let lam = _slang_Lambda_test_0(c);
+    int d = lam.operator()(2);
+}
+```
+
+## Environment Capturing
+
+If a lambda expression references a part of the environment variable either explicitly through a member or subscript operation
+or implicitly through `this` dereference, the entire object will be captured in the context. For example, given:
+
+```slang
+struct Composite
+{
+    int member1;
+    float member2;
+}
+
+void test()
+{
+    Composite c = {};
+    let lam = (int x) => x + c.member2;
+    lam(2);
+}
+```
+
+The generated `struct` type for the lambda expression will contain a member whose type is `Composite`, as in the following code:
+
+```slang
+void test()
+{
+    Composite c = {};
+    struct _slang_Lambda_test_0 : IFunc<int, int> {
+        // Lambda captures the entire object instead of just
+        // `member1`.
+        Composite c;
+        __init(Composite in_c) {
+            c = in_c;
```

Comment:
Are all types copyable/assignable in slang? Or if it's not, would trying to use its value in the lambda lead to a compile error?

Reply:
`RayQuery` and `HitObject` types are currently not copyable, and use of these in the capture will lead to a compile error.